nginx: add redirect for HTTP traffic

bit opened the pull request:
Once issue #942 is resolved, this should no longer be adding this condition again if it exists.

pde commented:
The core client team won't be able to review this until we've closed out the [1.0 for launch](https://github.com/letsencrypt/letsencrypt/issues?q=is%3Aopen+is%3Aissue+milestone%3A%221.0+for+launch%22) milestone.  But perhaps @diracdeltas or @erikrose might be able to get started on reviewing this in the meantime?

erikrose commented on letsencrypt-nginx/letsencrypt_nginx/parser.py:
```diff
@@ -215,7 +215,7 @@ def filedump(self, ext='tmp'):
                 with open(filename, 'w') as _file:
                     nginxparser.dump(tree, _file)
-            except IOError:
+            except IOError: # pragma: no cover
```
It seems a shame to no-cover these. We haven't done that for exception handlers elsewhere in the codebase. Perhaps we can find a way to exercise them without too much painful work.

erikrose commented on letsencrypt-nginx/letsencrypt_nginx/configurator.py:
```diff
@@ -362,6 +362,26 @@ def enhance(self, domain, enhancement, options=None):
+    def _enable_redirect(self, vhost, unused_options):
+        """Redirect all equivalent HTTP traffic to ssl_vhost.
+
+        Add rewrite directove to non https traffic
```
direct*i*ve :-)

erikrose commented on letsencrypt-nginx/letsencrypt_nginx/configurator.py:
```diff
@@ -362,6 +362,26 @@ def enhance(self, domain, enhancement, options=None):
+        .. note:: This function saves the configuration
+
+        :param vhost: Destination of traffic, an ssl enabled vhost
+        :type vhost: :class:`~letsencrypt_apache.obj.VirtualHost`
```
This probably isn't an apache vhost. :-)

erikrose commented on letsencrypt-nginx/letsencrypt_nginx/configurator.py:
```diff
@@ -362,6 +362,26 @@ def enhance(self, domain, enhancement, options=None):
+        :param unused_options: Not currently used
+        :type unused_options: Not Available
+        """
+        redirect_block = [[['if', '($scheme != "https")'],
```
This works fine, but a more efficient way would be `return 301 https://$host$uri;`. That avoids firing up the comparatively expensive regex machinery (and, for my reference, has been supported since 0.7.51, released in 2009).

We might be able to do even better in the future if we make a fresh vhost for the HTTP side:
```
server {
    server_name example.com;
    return 301 https://$host$request_uri;
}
```
That eliminates even the `if` but would require a bit of redesign.

erikrose commented:
Thanks for the patch! If you have a chance to touch up those few spots, I look forward to merging this in. nginx fans will thank you.

bit commented:
thanks for the feedback. 
Fixed typos and changed block to a `return 301`.

Looks like test coverage is ok without ignoring the IOError block,
so leaving that out from the first patch.

erikrose commented on letsencrypt-nginx/letsencrypt_nginx/configurator.py:
```diff
@@ -366,6 +366,26 @@ def enhance(self, domain, enhancement, options=None):
+        :type unused_options: Not Available
+        """
+        redirect_block = [[['if', '($scheme != "https")'],
+            [['return', '301 https://$host$uri']]
```
Any reason you went with $uri rather than $request_uri? $uri is normalized, changing `/foo/../a` to `/a` (which doesn't matter so much, since browsers do that themselves) but also having some [vaguely documented differences around index files](http://nginx.org/en/docs/http/ngx_http_core_module.html#var_uri) (which might matter, but I can't find a case where it does). My leaning is that $request_uri will have fewer surprises in store for us. (Their additional difference with regard to internal redirects shouldn't matter to us.)

bit replied:
$request_uri seams to be the better choice here.

erikrose commented:
@pde +1 for merge!

pde commented:
Great! Thanks @bit & @erikrose !
